examples/readonly-objects-and-fields/readonly_objects_and_fields.bal

```diff
@@ -0,0 +1,96 @@
+import ballerina/io;
+
+type Details record {|
+    int id;
+    string country;
+|};
+
+// An object with both read-only and non-read-only fields.
+type Employee object {
+    string department;
+    // `details` is a `readonly` field which cannot be updated once the value is created.
+    // `details` also expects an immutable value, and thus the effective contextually expected type for
+    // `details` is `Details & readonly`.
+    readonly Details details;
+
+    function init(string department, readonly & Details details) {
+        self.department = department;
+        self.details = details;
+    }
+};
+
+// An abstract object with non-readonly fields.
+type Controller abstract object {
+   int id;
+   string[] codes;
+
+   function getId() returns string;
+};
+
+// A non-abstract object with all read-only fields.
+type DefaultController object {
+    readonly int id;
+    readonly string[] codes;
+
+    function init(int id, readonly & string[] codes) {
+        self.id = id;
+        // The expected type for `codes` is `readonly & string[]`.
+        self.codes = codes;
+    }
+
+    function getId() returns string {
+        return string `Default: ${self.id}`;
+    }
+};
+
+// A non-abstract `readonly object`.
+type MainController readonly object {
+    int id;
+    string[] codes;
+
+    function init(int id, readonly & string[] codes) {
+        self.id = id;
+        // The expected type for `codes` is `readonly & string[]`.
+        self.codes = codes;
+    }
+
+    function getId() returns string {
+        return string `Main: ${self.id}`;
+    }
+};
+
+public function main() {
+
+    // Define an immutable `Details` value.
+    Details & readonly immutableDetails = {
+        id: 112233,
+        country: "Sri Lanka"
+    };
+
+    // Use `immutableDetails` as the value for the `readonly` `details` field in `Employee`.
+    Employee emp = new ("IT", immutableDetails);
+
+    Details details = emp.details;
+    // The `details` field of `emp` is immutable.
+    io:println("details is immutable: ", details.isReadOnly());
+
+    // A `readonly` intersection (`T & readonly`) can be used with an object only if it is an abstract object.
+    // A non-abstract object can be `readonly` either if all of its fields are `readonly` fields, or if the object
+    // is defined as a `readonly` object.
+    // In either case the fields cannot be updated once the value is created.
+    // Define a value of the non-abstract object type `DefaultController` with all `readonly` fields.
```

Comment:
This last sentence sorta sounds out of place with the rest of the content, since the other content are describing the concepts while this line is all of a sudden describing the line of code. Maybe can even do without the last line

Reply:
How about

```suggestion
    // Define a value of the non-abstract object type `DefaultController`, of which all the fields are `readonly` fields.
```

Reply:
No the sentence itself is ok. I meant the content before that explains the concept in a generic manner, without any reference to the code line. But then the last sentence is discussing the code, without connecting the previous content with the code line.

Reply:
What I was trying to do here was to emphasize that this is an object with only readonly fields. To show that we can directly use it as an immutable value.

Reply:
Will remove this since L84 mentions this anyway.